drakcore/drakcore/tools/ipt/ipt-blocks.cpp

```diff
@@ -0,0 +1,398 @@
+
+          if (cmd == PTW_CURRENT_CR3) {
+            current_cr3_ = data;
+          }
+          if (!show_drakvuf_) {
+            break;
+          }
+
+          switch (cmd) {
+            case PTW_CURRENT_CR3:
+              emit_event("drakvuf_cr3", to_hex(data));
+              break;
+            case PTW_CURRENT_TID:
+              emit_event("drakvuf_tid", data);
+              break;
+            case PTW_EVENT_ID:
+              emit_event("drakvuf_event", data);
+              break;
+            default:
+              emit_event("ptwrite", to_hex(event->variant.ptwrite.payload));
+          }
+        }
+        break;
+      case ptev_enabled:
+      case ptev_disabled:
+      case ptev_async_disabled:
+      case ptev_async_branch:
+      case ptev_paging:
+      case ptev_async_paging:
+      case ptev_overflow:
+      case ptev_exec_mode:
+      case ptev_tsx:
+      case ptev_stop:
+      case ptev_vmcs:
+      case ptev_async_vmcs:
+      case ptev_exstop:
+      case ptev_mwait:
+      case ptev_pwre:
+      case ptev_pwrx:
+      case ptev_mnt:
+      case ptev_tick:
+      case ptev_cbr:
+        break;
+    }
+  }
+
+ private:
+  void load_pt_mmap(const std::string &filename) {
+    if (mmap_buffer_ != nullptr) {
+      throw std::runtime_error("Trace alread mmaped");
+    }
+
+    int fd = open(filename.c_str(), O_CLOEXEC | O_RDONLY);
+    if (fd < 0) {
+      throw std::runtime_error("Cannot open " + filename);
+    }
+
+    struct stat file_stat;
+    int err = fstat(fd, &file_stat);
+    if (err < 0) {
+      throw std::runtime_error("Cannot stat PT file");
+    }
+
+    mmap_size_ = file_stat.st_size;
+
+    void *ptr = mmap(0, mmap_size_, PROT_READ, MAP_PRIVATE, fd, 0);
+    if (ptr == MAP_FAILED) {
+      throw std::runtime_error("Failed to mmap trace file");
+    }
+
+    mmap_buffer_ = ptr;
+    close(fd);
+
+    config_.begin = static_cast<uint8_t *>(ptr);
+    config_.end = static_cast<uint8_t *>(ptr) + mmap_size_;
+  }
+
+  void load_pt_vector(const std::string &filename) {
+    std::ifstream stream{filename, std::ios::binary};
+
+    stream.seekg(0, std::ios::end);
+    proc_trace_.reserve(stream.tellg());
+    stream.seekg(0, std::ios::beg);
+
+    proc_trace_.assign(std::istreambuf_iterator<char>{stream},
+                       std::istreambuf_iterator<char>{});
+
+    config_.begin = proc_trace_.data();
+    config_.end = config_.begin + proc_trace_.size();
+  }
+
+ public:
+  bool show_drakvuf_ = false;
+  bool enable_mmap_ = true;
+
+ private:
+  struct pt_config config_;
+
+  void *mmap_buffer_;
+  size_t mmap_size_;
+  std::vector<uint8_t> proc_trace_;
+
+  uint32_t current_cr3_;
+};
+
+int main(int argc, char *argv[]) {
+  auto err_console = spdlog::stderr_color_st("console");
+  spdlog::set_default_logger(err_console);
+
+  Decoder decoder{};
+  Image image{};
+
+  std::optional<std::string> pt_file;
+  std::optional<uint64_t> cr3_filter;
+
+  for (int i = 1; i < argc; i++) {
```

Comment:
Maybe a smol lib here? We already use spdlog, could use smth that would parse arguments.

Reply:
meh